Make `BeatmapSet.Files` non-nullable

peppy opened the pull request:
Split out from https://github.com/ppy/osu/pull/12860. Intended to improve sanity of the checks of this property for the purpose of knowing whether a beatmap is in a database fetched state or not.

- [x] Depends on https://github.com/ppy/osu/pull/12860

peppy committed:
Revert "Add temporary accounting for tests with null files"

This reverts commit e52c0a34f82a6b93c740b4c2d3d068bc3c52196a.

peppy committed:
Revert "Revert "Make `BeatmapSetInfo.Files` non-nullable""

This reverts commit 9c4f39e9681644d84b60828ee8ce613b6d81a89f.

peppy committed:
Revert "Revert "Fix editor tests failing due to empty files being specified""

This reverts commit 1af684c4b2cb41e45f473e90285f5ac5cc192d12.

peppy committed:
Remove no longer necessary conditional access

peppy committed:
Merge branch 'editor-exit-stability' into non-nullable-beatmapset-files

bdach committed:
Merge branch 'master' into non-nullable-beatmapset-files

smoogipoo commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -285,24 +282,20 @@ public void Save(BeatmapInfo info, IBeatmap beatmapContent, ISkin beatmapSkin =
 
-            if (beatmapInfo == null)
-                return DefaultBeatmap;
+            if (beatmapInfo.ID > 0 && previous != null && previous.BeatmapInfo?.ID == beatmapInfo.ID)
+                return previous;
```
Does this check really need to happen afterwards? If it's comparing by ID and returning the previous WB, the requery above is basically a no-op as far as I can tell.

peppy replied:
I think it is indeed redundant, will double check.

peppy replied:
I actually think this is not required with the `workingCache` logic directly following it, which should cover this exact scenario?

smoogipoo replied:
Yeah, I don't know why `previous` wasn't removed with the introduction of the cache. Removing this sounds good to me.

smoogipoo commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -241,12 +240,10 @@ protected override bool CheckLocalAvailability(BeatmapSetInfo model, IQueryable<
         {
             var setInfo = info.BeatmapSet;
 
-            Debug.Assert(setInfo.Files != null);
```
Are you sure about this removal? What's the case that this comes up?

smoogipoo replied:
Nevermind, was removed because it's "non-nullable". Though the property can be re-assigned...

peppy committed:
Remove `previous` consumption logic in `GetWorkingBeatmap`

This should not be required since the introduction of `workingCache`,
which does the same thing in a more global way.

smoogipoo commented:
There's an actual issue with the tests here.

peppy committed:
Merge branch 'master' into non-nullable-beatmapset-files

peppy committed:
Change check order around to ensure re-fetches which return no results don't nullref
